### PR TITLE
Autoload can_compress_scripts option

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/55270

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/includes/upgrade.php [hunk 1]
```diff
@@ -2293,6 +2297,28 @@ function upgrade_600() {
...
+ */
+function upgrade_630() {
+	global $wp_current_db_version;
+
+	if ( $wp_current_db_version < 55752 ) {
+		if ( ! is_multisite() ) {
+			// Replace non-autoload option can_compress_scripts with autoload option, see #55270
+			$can_compress_scripts = get_option( 'can_compress_scripts' ) ? 1 : 0;
+			delete_option( 'can_compress_scripts' );
+			add_option( 'can_compress_scripts', $can_compress_scripts );
```
Comment (mukeshpanchal27) [inline src/wp-admin/includes/upgrade.php]:
```suggestion
			update_option( 'can_compress_scripts', $can_compress_scripts, 'yes' );
```
Q: Instead of deleting and adding new option can we use `update_option`?

File: src/wp-admin/includes/upgrade.php [hunk 2]
```diff
@@ -2293,6 +2297,28 @@ function upgrade_600() {
 	}
 }
 
+
```
Comment (mukeshpanchal27) [inline src/wp-admin/includes/upgrade.php]:
```suggestion
```
Remove extra line break.

File: src/wp-admin/includes/upgrade.php [hunk 3]
```diff
@@ -2293,6 +2297,28 @@ function upgrade_600() {
...
+	if ( $wp_current_db_version < 55752 ) {
+		if ( ! is_multisite() ) {
+			// Replace non-autoload option can_compress_scripts with autoload option, see #55270
+			$can_compress_scripts = get_option( 'can_compress_scripts' ) ? 1 : 0;
+			delete_option( 'can_compress_scripts' );
+			add_option( 'can_compress_scripts', $can_compress_scripts );
+		}
+	}
+}
```
Comment (mukeshpanchal27) [inline src/wp-admin/includes/upgrade.php]:
```suggestion
/**
 * Executes changes made in WordPress 6.3.0.
 *
 * @ignore
 * @since 6.3.0
 */
function upgrade_630() {
	if ( ! is_multisite() ) {
		// Replace non-autoload option can_compress_scripts with autoload option, see #55270
		$can_compress_scripts = get_option( 'can_compress_scripts' ) ? 1 : 0;
		delete_option( 'can_compress_scripts' );
		add_option( 'can_compress_scripts', $can_compress_scripts );
	}
}
```

The `upgrade_630` function should only be executed when the `$wp_current_db_version` is less than 55752, so there's no need to check it here.

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @spacedmonkey, The changes look good to me. Left some feedback and questions.

- For fresh installations, the `can_compress_scripts` option is [deleted by default](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/includes/schema.php#L674). Do we need to remove it from there and add it in the default values?

File: src/wp-admin/includes/upgrade.php [hunk 4]
```diff
@@ -2293,6 +2297,28 @@ function upgrade_600() {
...
+	if ( $wp_current_db_version < 55752 ) {
+		if ( ! is_multisite() ) {
+			// Replace non-autoload option can_compress_scripts with autoload option, see #55270
+			$can_compress_scripts = get_option( 'can_compress_scripts', false );
+			if ( false !== $can_compress_scripts ) {
+				update_option( 'can_compress_scripts', $can_compress_scripts, 'yes' );
+			}
+		}
+	}
+}
```
Comment (mukeshpanchal27) [inline src/wp-admin/includes/upgrade.php]:
```suggestion
/**
 * Executes changes made in WordPress 6.3.0.
 *
 * @ignore
 * @since 6.3.0
 */
function upgrade_630() {
	if ( ! is_multisite() ) {
		// Replace non-autoload option can_compress_scripts with autoload option, see #55270
		$can_compress_scripts = get_option( 'can_compress_scripts', false );
		if ( false !== $can_compress_scripts ) {
			update_option( 'can_compress_scripts', $can_compress_scripts, 'yes' );
		}
	}
}
```

The `upgrade_630` function should only be executed when the `$wp_current_db_version` is less than 55752, so there's no need to check it here.

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks, @spacedmonkey, for the update. I approve the PR, but there is one minor review that needs to be addressed.

File: src/wp-admin/includes/upgrade.php (hunk 4 above)
Comment (spacedmonkey) [reply src/wp-admin/includes/upgrade.php]: @mukeshpanchal27 Compare this to other functions. https://github.com/WordPress/wordpress-develop/blob/a10e2bcf05309b814d7aba32281d24fe30eaf65b/src/wp-admin/includes/upgrade.php#L2266-L2278

This is how these work, so I going to leave this.

Comment (mukeshpanchal27) [reply src/wp-admin/includes/upgrade.php]: Yes but i think it wroth to check same condition again in function 🤔

Comment (spacedmonkey) [reply src/wp-admin/includes/upgrade.php]: > Yes but i think it wroth to check same condition again in function

Can you explain this?

Comment (mukeshpanchal27) [reply src/wp-admin/includes/upgrade.php]:
```
if ( $wp_current_db_version < 55752 ) {
	upgrade_630();
}
```

The `upgrade_630()` only call if db version is less then 55752 `if ( $wp_current_db_version < 55752 ) {` right?

```
function upgrade_630() {
	global $wp_current_db_version;
	if ( $wp_current_db_version < 55752 ) {
......
```

Again in `upgrade_630()` we check same condition `if ( $wp_current_db_version < 55752 ) {`.

Comment (spacedmonkey) [reply src/wp-admin/includes/upgrade.php]: That is what the other upgrade functions do. I don't see why this one would be any different.

File: src/wp-admin/includes/ajax-actions.php [hunk 5]
```diff
@@ -188,7 +188,11 @@ function wp_ajax_wp_compression_test() {
 	}
 
 	if ( ini_get( 'zlib.output_compression' ) || 'ob_gzhandler' === ini_get( 'output_handler' ) ) {
-		update_site_option( 'can_compress_scripts', 0 );
+		if ( is_multisite() ) {
+			update_site_option( 'can_compress_scripts', 0 );
+		} else {
+			update_option( 'can_compress_scripts', 0 );
+		}
```
Comment (felixarntz) [inline src/wp-admin/includes/ajax-actions.php]: This change is because `update_site_option()` does not allow control over the `$autoload` parameter when calling `update_option()`, is that right?

I think it would be better to explicitly pass "yes" here so that it's clear why the change is here. For someone that doesn't have the context, that wouldn't be obvious. Adding a comment in addition would be even better.
```suggestion
		// Use `update_option()` on single site to mark the option for autoloading.
		if ( is_multisite() ) {
			update_site_option( 'can_compress_scripts', 0 );
		} else {
			update_option( 'can_compress_scripts', 0, 'yes' );
		}
```

File: src/wp-admin/includes/ajax-actions.php [hunk 6]
```diff
@@ -222,10 +226,18 @@ function wp_ajax_wp_compression_test() {
 			wp_die();
 		} elseif ( 'no' === $_GET['test'] ) {
 			check_ajax_referer( 'update_can_compress_scripts' );
-			update_site_option( 'can_compress_scripts', 0 );
+			if ( is_multisite() ) {
+				update_site_option( 'can_compress_scripts', 0 );
+			} else {
+				update_option( 'can_compress_scripts', 0 );
+			}
```
Comment (felixarntz) [inline src/wp-admin/includes/ajax-actions.php]: See above.
```suggestion
			// Use `update_option()` on single site to mark the option for autoloading.
			if ( is_multisite() ) {
				update_site_option( 'can_compress_scripts', 0 );
			} else {
				update_option( 'can_compress_scripts', 0, 'yes' );
			}
```

File: src/wp-admin/includes/ajax-actions.php [hunk 7]
```diff
@@ -222,10 +226,18 @@ function wp_ajax_wp_compression_test() {
...
 		} elseif ( 'yes' === $_GET['test'] ) {
 			check_ajax_referer( 'update_can_compress_scripts' );
-			update_site_option( 'can_compress_scripts', 1 );
+			if ( is_multisite() ) {
+				update_site_option( 'can_compress_scripts', 1 );
+			} else {
+				update_option( 'can_compress_scripts', 1 );
+			}
```
Comment (felixarntz) [inline src/wp-admin/includes/ajax-actions.php]:
```suggestion
			// Use `update_option()` on single site to mark the option for autoloading.
			if ( is_multisite() ) {
				update_site_option( 'can_compress_scripts', 1 );
			} else {
				update_option( 'can_compress_scripts', 1, 'yes' );
			}
```

File: src/wp-admin/includes/upgrade.php (hunk 4 above)
Comment (felixarntz) [reply src/wp-admin/includes/upgrade.php]: I agree with @spacedmonkey here, while it seems indeed pointless, that's what all other WP core upgrade functions do, have the check both around the function and in the function. There may be some reason for that which we're missing, so I'd rather be safe and stick to that practice.

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey The code looks good to me, though I have one small suggestion to make the change more obvious. For someone looking at the code later, we should clarify.

File: src/wp-admin/includes/upgrade.php [hunk 8]
```diff
@@ -2293,6 +2297,28 @@ function upgrade_600() {
...
+ */
+function upgrade_630() {
+	global $wp_current_db_version;
+
+	if ( $wp_current_db_version < 55752 ) {
+		if ( ! is_multisite() ) {
+			// Replace non-autoload option can_compress_scripts with autoload option, see #55270
+			$can_compress_scripts = get_option( 'can_compress_scripts', false );
+			if ( false !== $can_compress_scripts ) {
+				update_option( 'can_compress_scripts', $can_compress_scripts, 'yes' );
```
Comment (spacedmonkey) [inline src/wp-admin/includes/upgrade.php]:
```suggestion
				delete_option( 'can_compress_scripts' );
				add_option( 'can_compress_scripts', $can_compress_scripts, 'yes' );
```

File: src/wp-admin/includes/upgrade.php (hunk 4 above)
Comment (mukeshpanchal27) [reply src/wp-admin/includes/upgrade.php]: Make sense. Thanks!

**Review by felixarntz — APPROVED** [review]
Thanks @spacedmonkey!

**Comment by spacedmonkey** [comment]
Committed https://core.trac.wordpress.org/changeset/55854
